src/jit/lsra.cpp

```diff
@@ -5005,28 +5111,87 @@ LinearScan::allocateBusyReg(Interval *current, RefPosition *refPosition)
         }
-#endif // DEBUG
+        else
+        // the below if-stmt is associated with this else
+#endif
+        if (recentAssignedRefWeight < farthestRefPosWeight)
+        {
+            isBetterLocation = true;
+        }
+        else
+        {
+            // This would mean the weight of spill ref position we found so far is equal
+            // to the weight of the ref position that is being evaluated.  In this case
+            // we prefer to spill ref position whose distance to its next reference is
+            // the farthest.
+            assert(recentAssignedRefWeight == farthestRefPosWeight);
+
+            // If allocation is optional, the first spill candidate selected
+            // will be based on weight alone. After we have found a spill
+            // candidate whose weight is less than the 'refPosition', we will 
+            // consider farthest distance when there is a tie in weights.
+            // This is to ensure that we don't spill a ref position whose
+            // weight is equal to weight of 'refPosition'.
+            if (allocationOptional && farthestRefPhysRegRecord == nullptr)
+            {
+                isBetterLocation = false;
+            }
+            else
+            {
+                isBetterLocation = (nextLocation > farthestLocation);
+            }
+        }
+
         if (isBetterLocation)
         {
             farthestLocation = nextLocation;
             farthestRefPhysRegRecord = physRegRecord;
+            farthestRefPosWeight = recentAssignedRefWeight;
         }
     }
-    assert(farthestRefPhysRegRecord != nullptr &&
-           (farthestLocation > refLocation || refPosition->isFixedRegRef));
-    foundReg = farthestRefPhysRegRecord->regNum;
-    unassignPhysReg(farthestRefPhysRegRecord, farthestRefPhysRegRecord->assignedInterval->recentRefPosition);
-    assignPhysReg(farthestRefPhysRegRecord, current);
 
-    refPosition->registerAssignment = genRegMask(foundReg);
+#if DEBUG
+    if (allocationOptional)
+    {
+        // There may not be a spill candidate or if one is found
+        // its weight must be less than the weight of 'refPosition'
+        assert((farthestRefPhysRegRecord == nullptr) ||
+               (farthestRefPosWeight < getWeight(refPosition)));
+    }
+    else 
+    {
+        // Must have found a spill candidate.
+        assert((farthestRefPhysRegRecord != nullptr) &&
+               (farthestLocation > refLocation || refPosition->isFixedRegRef));
+    }
+#endif
+
+    if (farthestRefPhysRegRecord != nullptr)
+    {
+        foundReg = farthestRefPhysRegRecord->regNum;
+        unassignPhysReg(farthestRefPhysRegRecord, farthestRefPhysRegRecord->assignedInterval->recentRefPosition);
+        assignPhysReg(farthestRefPhysRegRecord, current);
+        refPosition->registerAssignment = genRegMask(foundReg);
+    }
+    else
+    {
+        // This can happen only if allocating a reg is optional.
+        assert(allocationOptional);
```

Comment:
I am not sure I understood why you said it is redundant.  After looping through all busy registers, if we haven't found a spill candidate implies, it must be marked as reg optional.  Otherwise, the loop should always ensure that it finds a spill candidate.  This is essentially asserting here the condition that should be ensured by allocateBusyReg() loop.